sass/test.hack

```diff
@@ -0,0 +1,28 @@
+// This class is readonly. Had we put in a setter for $this->t, we could not
+// use covariance. e.g., if we had function setMe(T $x), you would get this
+// cov.php:9:25,25: Illegal usage of a covariant type parameter (Typing[4120])
+//   cov.php:7:10,10: This is where the parameter was declared as covariant (+)
+//   cov.php:9:25,25: Function parameters are contravariant
+class C<+T> {
+  public function __construct(private T $t) {}
+}
+
+class Animal {}
+class Cat extends Animal {}
+
+function f(C<Animal> $p1): void {
+  \var_dump($p1);
+}
+
+function g(vec<Animal> $p1): void {
+  \var_dump($p1);
+}
+
+<<__EntryPoint>>
+function run(): void {
+  f(new C(new Animal()));
+  f(new C(new Cat())); // accepted
+
+  g(vec[new Animal(), new Animal()]);
+  g(vec[new Cat(), new Cat(), new Animal()]); // arrays are covariant
+}
```

Comment:
I'm guessing you didn't mean to commit this? It doesn't look like this file is used anywhere.

If you want to demonstrate covariance of arrays, the easiest way is probably to contrast it with `Vector`, where we can treat a `vec<Dog>` as a `vec<Animal>` but we can't treat as `Vector<Dog>` as a `Vector<Animal>` and use it as a `Vector<Dog>` again after.

```
class Animal {}
class Dog extends Animal {
  public function bark(): void {}
}
class Cat extends Animal {}

function add_item_vec(vec<Animal> $items): vec<Animal> {
  // vec is a value type, so this creates a copy.
  $items[0] = new Cat();
  return $items;
}

function add_item_vector(Vector<Animal> $items): Vector<Animal> {
  // Vector is a reference type, so this modifies the original.
  $items[0] = new Cat();
  return $items;
}

function demo(): void {
  $dogs = vec[new Dog()];
  add_item_vec($dogs);
  $dogs[0]->bark(); // this is fine because we didn't modify $dogs.

  $dogs2 = Vector {new Dog()};
  add_item_vector($dogs2);
  $dogs2[0]->bark(); // hack correctly complains $dogs2 is now a Vector<Animal>.
}
```

Reply:
ughh again with the test file, sorry